### PR TITLE
EOS-22448 Accept Actuator Request Across Nodes

### PR DESCRIPTION
**Pull request by jjxsg**
## Problem Statement:

Accept Actuator Request Across Nodes
## Solution Design:

Broadcast request to each node and identify intended recipient based on node identifier. Broadcasting is achieved by assigning different consumer group to each consumer node. 
## Coding:

* [x] Did you follow coding standard and verified it with flake8? 

* [x] Did your address all Codacy issues?

## Testing:

* [ ] Unittest updated and all unittests are passing?

* [ ] Sanity/Self tests are updated (if applicable)?

* [ ] Manual testing done covering happy path and non-happy path?

## Integration:

* [x] If there is any interface change, did you communicate it to other components?
* [x] If there is any interface change, other component gate gatekeepers are ready to accept the change?
* [ ] Did you complete deployment test if applicable?

## PR checklist:
* [x] Did you add Jira number to the PR?

* [x] DCO and cla-signed?

### CODE REVIEW

File: low-level/framework/messaging/ingress_processor.py [hunk 1]
```diff
@@ -243,9 +257,7 @@ def _read_config(self):
         self._consumer_id = Conf.get(SSPL_CONF,
                                      f"{self.PROCESSOR}>{self.CONSUMER_ID}",
                                      'sspl_actuator')
-        self._consumer_group = Conf.get(SSPL_CONF,
-                                        f"{self.PROCESSOR}>{self.CONSUMER_GROUP}",
-                                        'cortx_monitor')
+        self._consumer_group = socket.getfqdn()
```
Comment (jjxsg) [inline low-level/framework/messaging/ingress_processor.py:260]: Broadcasting requires all node to be in different consumer groups. As of now I am using fqdn. Please suggest if something else should be used or it should be extracted from config. @cdeshmukh @sandeepsgt @shriya-deshmukh

File: low-level/framework/messaging/ingress_processor.py [hunk 2]
```diff
@@ -193,9 +195,21 @@ def _process_msg(self, body):
             self._check_debug(message)
             self._log_debug("_process_msg, ingressMsg: %s" % ingressMsg)
 
-            self._send_to_msg_handler(msgType, message, uuid)
+            # Compare hostname from the request to determine
+            # if request is meant for the current node
+            node_identifier = ingressMsg.get("node_identifier")
```
Comment (sandeepsgt) [inline low-level/framework/messaging/ingress_processor.py:200]: So CSM will add node_identifier key with hostname as value ?

File: low-level/framework/messaging/ingress_processor.py (hunk 1 above)
Comment (sandeepsgt) [reply low-level/framework/messaging/ingress_processor.py]: cortx_monitor_<node_id>

we used to have this for rabbitmq routing key, which served similar purpose. Biding messages to node
So, I think you can use above

File: low-level/framework/messaging/ingress_processor.py [hunk 3]
```diff
@@ -149,7 +149,7 @@ def _process_msg(self, body):
 
         ingressMsg = {}
         uuid = None
-        hostname = socket.getfqdn()
+        hostname = socket.gethostname()
```
Comment (cdeshmukh) [inline low-level/framework/messaging/ingress_processor.py:152]: so check for hostname already exists, which is serving this usecase already

File: low-level/framework/messaging/ingress_processor.py [hunk 4]
```diff
@@ -257,7 +257,7 @@ def _read_config(self):
         self._consumer_id = Conf.get(SSPL_CONF,
                                      f"{self.PROCESSOR}>{self.CONSUMER_ID}",
                                      'sspl_actuator')
-        self._consumer_group = socket.getfqdn()
+        self._consumer_group = socket.gethostname()
```
Comment (cdeshmukh) [inline low-level/framework/messaging/ingress_processor.py:260]: why this needed ?

File: low-level/framework/messaging/ingress_processor.py (hunk 2 above)
Comment (jjxsg) [reply low-level/framework/messaging/ingress_processor.py]: We will need to communicate this to the end users (HA, CSM). I am thinking of some way to keep in backward compatible so if someone does not include the `node_identifier`, the request should proceed in old way.

File: low-level/framework/messaging/ingress_processor.py (hunk 4 above)
Comment (jjxsg) [reply low-level/framework/messaging/ingress_processor.py:260]: In order to broadcast request to each node we need to assign a unique consumer group to each consumer node. We can use any value which is unique for a given node across cluster. Like machine-id, hostname etc.

File: low-level/framework/messaging/ingress_processor.py (hunk 3 above)
Comment (jjxsg) [reply low-level/framework/messaging/ingress_processor.py:152]: Sorry @cdeshmukh, I didn't get you, what do you mean by hostname already exists?

File: low-level/framework/messaging/ingress_processor.py (hunk 1 above)
Comment (jjxsg) [reply low-level/framework/messaging/ingress_processor.py]: I initially tried with node_id but we dont have node_id defined in SSPL_CONF so it takes the default value as `SN01`
```
self._node_id = Conf.get(SSPL_CONF,f"{CLUSTER}>{SRVNODE}>{self.NODE_ID_KEY}", 'SN01')
```

Comment (sandeepsgt) [reply low-level/framework/messaging/ingress_processor.py]: node_id is in global conf, not in sspl conf

Comment (jjxsg) [reply low-level/framework/messaging/ingress_processor.py]: I have updated `consumer_group` to `consumer_group_prefix` in `sspl.conf` which will be prefixed with Node ID.

File: low-level/files/opt/seagate/sspl/conf/sspl.conf.LR2.yaml [hunk 5]
```diff
@@ -46,7 +46,7 @@ SSPL_LL_SETTING:
 
 INGRESSPROCESSOR:
    consumer_id: sspl_actuator
-   consumer_group: cortx_monitor
+   consumer_group_prefix: cortx_monitor_
```
Comment (cdeshmukh) [inline low-level/files/opt/seagate/sspl/conf/sspl.conf.LR2.yaml:49]: _ added to append node-id ? append as "-node-id" only then, keeping it clean here "cortx_monitor"

File: low-level/framework/messaging/ingress_processor.py [hunk 6]
```diff
@@ -185,31 +185,44 @@ def _process_msg(self, body):
                     "IngressProcessor, Authentication failed on message: %s" % ingressMsg)
                 return
 
+            # Check for debugging being activated in the message header
+            self._check_debug(message)
```
Comment (cdeshmukh) [inline low-level/framework/messaging/ingress_processor.py:189]: use logging function like everywhere else, this was legacy way
logger.debug doesnt suffice ?

File: low-level/framework/messaging/ingress_processor.py [hunk 7]
```diff
@@ -185,31 +185,44 @@ def _process_msg(self, body):
...
             if message.get("actuator_request_type") is not None:
                 msgType = message.get("actuator_request_type")
 
                 # Validate against the actuator schema
                 validate(ingressMsg, self._actuator_schema)
+                # Compare hostname from the request to determine
+                # if request is meant for the current node
+                target_node_id = message.get("target_node_id")
+                if target_node_id is None:
+                    logger.warning(
```
Comment (cdeshmukh) [inline low-level/framework/messaging/ingress_processor.py:202]: logger.warning(
                        "Required attribute target_node_id is missing "
                        "from actuator request with %s, IGNORING!!".format(request))

request needs to be deterministic from logs also

File: low-level/framework/messaging/ingress_processor.py [hunk 8]
```diff
@@ -185,31 +185,44 @@ def _process_msg(self, body):
...
+                        "Required attribute target_node_id is missing "
+                        "from actuator request")
+                    return
+                elif target_node_id == self._node_id:
+                    self._send_to_msg_handler(msgType, message, uuid)
+                else:
+                    logger.info(
+                        "Node identifier mismatch, actuator request rejected.")
```
Comment (cdeshmukh) [inline low-level/framework/messaging/ingress_processor.py:210]: logger.debug
rejected -> ignored

File: low-level/framework/messaging/ingress_processor.py [hunk 9]
```diff
@@ -185,31 +185,44 @@ def _process_msg(self, body):
...
+                    return
 
             elif message.get("sensor_request_type") is not None:
                 msgType = message.get("sensor_request_type")
 
                 # Validate against the sensor schema
                 validate(ingressMsg, self._sensor_schema)
+                self._send_to_msg_handler(msgType, message, uuid)
```
Comment (cdeshmukh) [inline low-level/framework/messaging/ingress_processor.py:218]: why this change? relating to sensor rt !

File: low-level/framework/messaging/ingress_processor.py [hunk 10]
```diff
@@ -185,31 +185,44 @@ def _process_msg(self, body):
...
                 return
 
-            # Check for debugging being activated in the message header
-            self._check_debug(message)
-            self._log_debug("_process_msg, ingressMsg: %s" % ingressMsg)
-
-            self._send_to_msg_handler(msgType, message, uuid)
-
-        except Exception as ex:
+        except Exception:
```
Comment (cdeshmukh) [inline low-level/framework/messaging/ingress_processor.py:225]: we should always capture exception in a var and printed / passed

File: low-level/framework/messaging/ingress_processor.py [hunk 11]
```diff
@@ -248,18 +261,17 @@ def _send_to_msg_handler(self, msgType, message, uuid):
...
         """Read config for messaging bus."""
         # Make methods locally available
-        self._node_id = Conf.get(SSPL_CONF,
-                                 f"{CLUSTER}>{SRVNODE}>{self.NODE_ID_KEY}",
-                                 'SN01')
+        self._node_id = Conf.get(GLOBAL_CONF, NODE_ID_KEY, 'SN01')
+        self._consumer_group_prefix = Conf.get(
+            SSPL_CONF, f"{self.PROCESSOR}>{self.CONSUMER_GROUP}",
+            'cortx_monitor')
+        self._consumer_group = self._consumer_group_prefix + str(self._node_id)
```
Comment (cdeshmukh) [inline low-level/framework/messaging/ingress_processor.py:271]: here we can append with - and then nodeid

File: low-level/json_msgs/schemas/actuators/SSPL-LL_Actuator_Request.json [hunk 12]
```diff
@@ -73,7 +73,10 @@
 						"required": false
 					}
 				},
-
+				"target_node_id": {
```
Comment (cdeshmukh) [inline low-level/json_msgs/schemas/actuators/SSPL-LL_Actuator_Request.json:76]: this is going to fail any new existing actuator requests, so double confirm with CSM again

File: sspl_test/alerts/os/test_systemd_service_actuator.py [hunk 13]
```diff
@@ -123,7 +123,7 @@ def test_systemd_service_invalid_request(args):
     assert (specific_info[0].get("error_msg") is not None)
 
 
-def service_actuator_request(service_name, action):
+def service_actuator_request(service_name, action, target_node_id="SN01"):
```
Comment (cdeshmukh) [inline sspl_test/alerts/os/test_systemd_service_actuator.py:126]: pick default value from constants

File: low-level/files/opt/seagate/sspl/conf/sspl.conf.LR2.yaml (hunk 5 above)
Comment (jjxsg) [reply low-level/files/opt/seagate/sspl/conf/sspl.conf.LR2.yaml]: Done

File: low-level/framework/messaging/ingress_processor.py (hunk 6 above)
Comment (jjxsg) [reply low-level/framework/messaging/ingress_processor.py:189]: Changed to `logger.debug()`
We have an attribute `debug_enabled` in actuator request to indicate whether to enable debug logging or not.
`_check_debug()` function checks and enable that attribute.
Sample actuator request:
```
{
  "title": "SSPL-LL Actuator Request",
  "description": "Seagate Storage Platform Library - Low Level - Actuator Request",
  "username": "sspl-ll",
  "expires": 3600,
  "signature": "None",
  "time": "2018-07-31 04:08:04.071170",
  "message": {
    "sspl_ll_debug": {
      "debug_component": "sensor",
      "debug_enabled": true
    },
    "sspl_ll_msg_header": {
      "msg_version": "1.0.0",
      "uuid": "16476007-a739-4785-b5c7-f3de189cdf9d",
      "schema_version": "1.0.0",
      "sspl_version": "1.0.0"
    },
"request_path": {
      "site_id": "1",
      "rack_id": "1",
      "node_id": "1"
    },
    "response_dest": {},
    "target_node_id":"1",
    "actuator_request_type": {
      "service_controller": {
      "service_request": "status",
      "service_name": "kafka.service"
       }
    }
  }
}
```

File: low-level/framework/messaging/ingress_processor.py (hunk 7 above)
Comment (jjxsg) [reply low-level/framework/messaging/ingress_processor.py:202]: Done

File: low-level/framework/messaging/ingress_processor.py (hunk 8 above)
Comment (jjxsg) [reply low-level/framework/messaging/ingress_processor.py]: Done

File: low-level/framework/messaging/ingress_processor.py (hunk 9 above)
Comment (jjxsg) [reply low-level/framework/messaging/ingress_processor.py:218]: I have just rearranged the code. Earlier there were no checks so there was just one call of `_send_to_msg_handler()` for both sensor and actuator requests. Now since we have `target_node_id` for actuator, I added separate call for sensor requests.

File: low-level/framework/messaging/ingress_processor.py (hunk 10 above)
Comment (jjxsg) [reply low-level/framework/messaging/ingress_processor.py]: Actually, it was picked up by Codacy as unused variable. I will add it to the log statement.

File: low-level/framework/messaging/ingress_processor.py (hunk 11 above)
Comment (jjxsg) [reply low-level/framework/messaging/ingress_processor.py]: Done

File: low-level/json_msgs/schemas/actuators/SSPL-LL_Actuator_Request.json (hunk 12 above)
Comment (jjxsg) [reply low-level/json_msgs/schemas/actuators/SSPL-LL_Actuator_Request.json:76]: I have asked @pawankumarsrivastava to double check and he confirmed. I will create a ticket for the same.

File: sspl_test/alerts/os/test_systemd_service_actuator.py (hunk 13 above)
Comment (jjxsg) [reply sspl_test/alerts/os/test_systemd_service_actuator.py]: Done

**Review by sandeepsgt — APPROVED** [review]
This again is broadcast approach, but in arch discussion we had decided to use node specific queues.
There was some confusion in using terminologies, but main idea was to avoid broadcasting.

File: sspl_test/alerts/node/test_node_disk_actuator.py [hunk 14]
```diff
@@ -27,7 +29,8 @@ def test_node_disk_module_actuator(agrs):
     check_sspl_ll_is_running()
     instance_id = "*"
     resource_type = "node:fru:disk"
-    send_node_controller_message_request(UUID, "NDHW:%s" % resource_type, instance_id)
+    target_node_id = Conf.get(SSPL_TEST_CONF, NODE_ID_KEY, DEFAULT_NODE_ID)
```
Comment (cdeshmukh) [inline sspl_test/alerts/node/test_node_disk_actuator.py:32]: build this target_node_id in common.py and feed that to send_node_controller_message_request() in same file, by default. that way you wont need changes in all this test code

Comment (jjxsg) [reply sspl_test/alerts/node/test_node_disk_actuator.py]: Done

File: low-level/framework/messaging/ingress_processor.py (hunk 7 above)
Comment (jjxsg) [reply low-level/framework/messaging/ingress_processor.py:202]: I have removed %s and with from the statement. In egress processer, when request message is sent, entire message is printed in log at debug level, so kept level debug here as well. 
And if `target_node_id` is missing, it should be a warning.

File: low-level/framework/messaging/ingress_processor.py (hunk 6 above)
Comment (jjxsg) [reply low-level/framework/messaging/ingress_processor.py:189]: > sample request looks fine, except below part. target_node_id is the only new field, mandatory one. rest remains same i believe

Yes, only `target_node_id` is added, rest is same.
>basically no Low level reference should be there in any code atleast

`low level` is not present in code, removed from the comment here as well.

Comment (jjxsg) [reply low-level/framework/messaging/ingress_processor.py:189]: >Yeah had seen that, we dont use that any more in new requests atleast, all log levels managed by set log level config & not through request

Should I remove `_check_debug(message)` function?

File: sspl_test/alerts/os/test_systemd_service_actuator.py [hunk 15]
```diff
@@ -21,12 +21,19 @@
...
 RESOURCE_TYPE = "node:sw:os:service"
 
 def init(args):
     pass
 
+def get_current_node_id():
+    """Get current node id."""
+    node_id = Conf.get(GLOBAL_CONF, NODE_ID_KEY, DEFAULT_NODE_ID)
+    return node_id
+
```
Comment (sumedhak27) [inline sspl_test/alerts/os/test_systemd_service_actuator.py:36]: Hi, why are we redefining this function?? you defined it already in sspl_test/common.py, right?

File: low-level/framework/messaging/ingress_processor.py (hunk 6 above)
Comment (cdeshmukh) [reply low-level/framework/messaging/ingress_processor.py:189]: remove its usage from here, thats enough. rest of the cleanup we will take up separately

File: low-level/framework/messaging/ingress_processor.py (hunk 10 above)
Comment (jjxsg) [reply low-level/framework/messaging/ingress_processor.py]: Added exception variable in log statement.

File: sspl_test/alerts/os/test_systemd_service_actuator.py (hunk 15 above)
Comment (jjxsg) [reply sspl_test/alerts/os/test_systemd_service_actuator.py]: Yes, removed it from all other file and importing it from common.py

File: low-level/framework/messaging/ingress_processor.py (hunk 6 above)
Comment (jjxsg) [reply low-level/framework/messaging/ingress_processor.py]: Removed `_check_debug(message)`

**Review by sumedhak27 — APPROVED** [review]

File: sspl_test/alerts/os/test_systemd_service_actuator.py [hunk 16]
```diff
@@ -148,6 +150,7 @@ def service_actuator_request(service_name, action):
                         "node_id": "1"
                     },
                     "response_dest": {},
+                    "target_node_id": node_id,
```
Comment (cdeshmukh) [inline sspl_test/alerts/os/test_systemd_service_actuator.py:153]: so this code tests actuator request on same now. which is fine

Comment (jjxsg) [reply sspl_test/alerts/os/test_systemd_service_actuator.py]: Yes, we will need to make changes here later as per the 3-node sanity plan.

File: sspl_test/alerts/realstor/test_real_stor_for_platform_sensor.py [hunk 17]
```diff
@@ -214,6 +216,7 @@ def enclosure_sensor_message_request(resource_type, resource_id):
                 "node_id": "1"
             },
             "response_dest": {},
+            "target_node_id": node_id,
```
[the remark this answers is not in the document]
Comment (jjxsg) [reply sspl_test/alerts/realstor/test_real_stor_for_platform_sensor.py]: added direct call to `get_current_node_id()` and removed `node_id` variable

File: low-level/framework/messaging/ingress_processor.py (hunk 7 above)
Comment (jjxsg) [reply low-level/framework/messaging/ingress_processor.py:200]: Added attributes from `actuator_request_type` in warning message.

File: sspl_test/messaging/ingress_processor_tests.py [hunk 18]
```diff
@@ -219,16 +220,18 @@ def _process_msg(self, body):
             logger.exception(
                 "_process_msg unrecognized message: %r" % ingressMsg)
 
-    def _read_config(self):
+    def _init_config(self):
         """Configure the messaging exchange with defaults available."""
         # Make methods locally available
-        self._node_id = Conf.get(SSPL_TEST_CONF, NODE_ID_KEY, 'SN01')
+        self._node_id = Conf.get(GLOBAL_CONF, NODE_ID_KEY, DEFAULT_NODE_ID)
```
Comment (shriya-deshmukh) [inline sspl_test/messaging/ingress_processor_tests.py:226]: Here also you can use get_current_node_id().

**Review by shriya-deshmukh — APPROVED** [review]

File: sspl_test/messaging/ingress_processor_tests.py (hunk 18 above)
Comment (jjxsg) [reply sspl_test/messaging/ingress_processor_tests.py:225]: Tried this earlier but it was failing with import error. 
```
File "/opt/seagate/cortx/sspl/sspl_test/run_test.py", line 33, in <module>
from common import (TestFailed, init_messaging_msg_processors,
File "/opt/seagate/cortx/sspl/sspl_test/common.py", line 37, in <module>
from messaging.ingress_processor_tests import IngressProcessorTests 
File "/opt/seagate/cortx/sspl/sspl_test/messaging/ingress_processor_tests.py", line 37, in <module>
from common import get_current_node_id
ImportError: cannot import name 'get_current_node_id'
```

File: low-level/json_msgs/schemas/actuators/SSPL-LL_Actuator_Request.json (hunk 12 above)
Comment (cdeshmukh) [reply low-level/json_msgs/schemas/actuators/SSPL-LL_Actuator_Request.json:76]: ok

**Review by cdeshmukh — APPROVED** [review]

**Review by sandeepsgt — APPROVED** [review]

**Review by shriya-deshmukh — APPROVED** [review]

File: low-level/sensors/impl/generic/cpu_fault_sensor.py [hunk 19]
```diff
@@ -218,7 +218,7 @@ def _create_json_message(self, cpu, alert_type):
         epoch_time = str(int(time.time()))
 
         alert_id = self._get_alert_id(epoch_time)
-        host_name = socket.gethostname()
+        host_name = socket.getfqdn()
```
Comment (cdeshmukh) [inline low-level/sensors/impl/generic/cpu_fault_sensor.py:221]: getfqdn() is changing the host_name in alerts to private nw interface name, which is not what we need

>
>>> import socket
>>> socket.getfqdn()
'srvnode-3'
>>> socket.gethostname()
'sm17-r18.pun.seagate.com'
>>>

**Review by cdeshmukh — COMMENTED** [review]
Cant change to getfqdn() as such in sensor responses as in CSM interfaces, hostname needs to reflect public data hostname, what gethostname() would give i believe

File: low-level/sensors/impl/generic/cpu_fault_sensor.py (hunk 19 above)
Comment (cdeshmukh) [reply low-level/sensors/impl/generic/cpu_fault_sensor.py:221]: There was a review comment by Ujjwal that we should use private hostname to link to these consumer groups we are introducing for Actuators, as public hostname may get changed in field. so in that getfqdn() is needed to get us private nw hostname, and not gethostname()
​
And obviously in sensor also we need to have same value. We will have to intimate CSM about this change, possibly by creating a jira ticket for them to check & assess

File: low-level/sensors/impl/generic/node_memory_fault.py [hunk 20]
```diff
@@ -204,7 +204,7 @@ def _create_json_message(self, alert_type):
         epoch_time = str(int(time.time()))
 
         alert_id = self._get_alert_id(epoch_time)
-        host_name = socket.gethostname()
+        host_name = socket.getfqdn()
```
[the remark this answers is not in the document]
Comment (jjxsg) [reply low-level/sensors/impl/generic/node_memory_fault.py]: I have added `OSUtils` class. Please check.

File: low-level/sensors/impl/centos_7/service_monitor.py [hunk 21]
```diff
@@ -499,7 +499,7 @@ def get_alert(cls, service, alert):
         return {
             "sensor_request_type": {
                 "service_status_alert": {
-                    "host_id": socket.getfqdn(),
+                    "host_id": OSUtils.get_fqdn(),
```
Comment (cdeshmukh) [inline low-level/sensors/impl/centos_7/service_monitor.py:502]: Using this way is better

**Review by cdeshmukh — APPROVED** [review]
